File: solutions/0000-0099/42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public static int trap(int[] a) {
        int[] l= new int[a.length];
        int[] r= new int[a.length];
        l[0]=a[0];
        for (int i = 1; i <a.length ; i++) {
            l[i]=Math.max(l[i-1],a[i]);
        }
       r[a.length-1]=a[a.length-1];
        for (int i = a.length-2; i >=0 ; i--) {
            r[i]=Math.max(r[i+1],a[i]);
        }

        int sum=0;

        for (int i = 1; i < a.length-1; i++) {
            sum+=Math.min(l[i],r[i])-a[i];
        }
        return sum;
    }
}

File: solutions/0000-0099/42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& nums) {
        
        int ans = 0;
        int n = nums.size();
        vector<int> left(n);
        vector<int> right(n);
        int lmaxi = 0;
        int rmaxi = 0;
        for(int i=0;i<nums.size();i++)
        {
            left[i] = lmaxi = max(lmaxi,nums[i]);
            right[n-1-i]  = rmaxi = max(rmaxi,nums[n-i-1]);
        }
        
        for(int i=0;i<nums.size();i++)
        {
            ans += min(left[i],right[i]) - nums[i];
        }
        return ans;
      
    }
};